src/Lorentz/Contracts/BaseDAO/Types.hs

```diff
@@ -217,7 +217,7 @@ data Storage (contractExtra :: Kind.Type) (proposalMetadata :: Kind.Type) = Stor
 
   , sExtra                     :: contractExtra
   , sProposals                 :: BigMap (ProposalKey proposalMetadata) (Proposal proposalMetadata)
-  , sProposalKeyListSortByDate :: [ProposalKey proposalMetadata] -- Newest first
+  , sProposalKeyListSortByDate :: Map Timestamp (ProposalKey proposalMetadata) -- Oldest first
```

Comment:
This MR is still WIP, currently allow `flush` to accepts an amount of proposals is done. I will work on adding `dropProposals` entrypoint and their tests next.

One thing that I need advice on is that, currently, I have to update `sProposalKeyListSortByDate` from a list to map of proposal created date and its key.

This is due to 2 main reasons:
- There is currently a bug in `handleProposalIsOver`: https://github.com/tqtezos/baseDAO/blob/master/src/Lorentz/Contracts/BaseDAO/Proposal.hs#L656-L663
  - I implemented this naively, this end up reversing the list. 
  Since we process all the proposals all at the same time, it was hard to detect this.

- Using a list is hard to modify. Since we can handle certain amount of a proposals now, using `Map` is much easier to update/remove the proposal key.

Hope to get your views on this @gromakovsky @Martoon-00 .

Reply:
I remember that in another project we also sometimes had to use maps for structures that were initially list-like - because of some Michelson limitations. So I think it's a good change if it simplifies things.

However, note that keeping `Map Timestamp a` may be tricky since `NOW` eventually returns the time of operation inclusion into a block, thus probability of collision is high. You will probably have to use `Map Timestamp [a]`. And if so, I wonder whether will we get any benefits from using Map. 

Another solution is to use numeric indices as keys and remember the indices of the first and last active proposals or something like that. And add an index in that map into `Proposal` datatype. This is gonna be even less pleasant than working with `Map Timestamp` though, but the most correct and flexible.

Could you elaborate on the second point: "Using a list is hard to modify"? AFAIU the first point is caused by the lack of tests on storage (#31) and is quite fixable. Regarding the second point, probably the following dareful plain would be possible to implement:
* Keep the list in "oldest goes first" order;
* Now we don't even need to limit the number of active proposals, because when it is high, `propose` will be the most expensive operation and thus the number of proposals will become naturally limited.
* `ensureProposalIsUnique` can probably be implemented via seeking by proposal hash in proposals map;
* `addProposal` will add to the end of the list;
* `flush` will do a series of `IF_CONS` to consume the list in direct order.

This plan probably has its own issues, but if everything is agreed it seems the simplest to implement to me.

What do you think?

Reply:
> probability of collision is high

I didn't think of this, and this could be a big issue.

> Using a list is hard to modify
This is pretty much due to having to copy to a new list when we add/remove an element form a list. Since it is actually linear, I guess it is not a performance issue. 

I'm good with your points, and I will make these changes instead of using `Map`.

Reply:
> This is pretty much due to having to copy to a new list when we add/remove an element form a list. It is simply a bit tricky, and I was worried about the performance. But in our case, it seems to be linear anyway, I guess that wouldn't be a performance problem.

Yeah, right. Well, if we would use `BigMap Index Proposal`, then we could really have `O(log N)` cost of `propose` and `O(n)` cost for `flush(n)`, though this would be trickier to support.

Reply:
The issue with `BigMap` is that I don't think we can loop through it using `ITER`.

Related to the collision, I'm thinking of using `Map (Timestamp, ProposalKey) ()`, this way we can guarantee that it has no collision, it should also be sorted with timestamp as well, since it seems in michelson, it compare the left of a pair first.

With this in mind, do you think using `Map` could be better? If not, I'll still go with List.

Reply:
> The issue with BigMap is that I don't think we can loop through it using ITER.

Yep, but if we store indices of the first and last active proposals, then in all cases we can iterate by indices, in theory.

> Map (Timestamp, ProposalKey) ()

And then `ITER` over it to get proposals in appropriate order? Right this should work now. I though that there is no way to get all elements with timestamp less than a given value from such a map, but apparently `ITER` suits that.

Okay, I think it's reasonable to use map, should be safer at least, and modifying it is indeed much simpler.